libraries/AP_Telemetry/AP_Telemetry_MQTT.cpp

```diff
@@ -0,0 +1,252 @@
+            printf("Failed to start connect, return code %d\n", rc);
+            MQTTHandle_error(rc);
+        }
+        connection_status = MQTT_CONNECTED;
+    }
+}
+
+void AP_Telemetry_MQTT::set_mqtt_server(const char* server)
+{
+    mqtt_server = server;
+}
+
+void AP_Telemetry_MQTT::set_mqtt_user(const char* user)
+{
+    conn_options.username = user;
+}
+
+void AP_Telemetry_MQTT::set_mqtt_password(const char* password)
+{
+    conn_options.password = password;
+}
+
+void AP_Telemetry_MQTT::send_log(const char* str)
+{
+    char topic[MAX_PAYLOAD];
+    if (send_log_flag == MQTT_SEND_LOG_ON) {
+        if (connection_status == MQTT_CONNECTED) {
+            sprintf(topic, "$ardupilot/copter/quad/log/%04d/location",
+                    mavlink_system.sysid);
+            send_message(str, topic);
+        }
+    }
+}
+
+void AP_Telemetry_MQTT::send_message(const char *str, const char *topic)
+{
+    int rc;
+    MQTTAsync_message pubmsg = MQTTAsync_message_initializer;
+    char tmp[MAX_PAYLOAD];
+
+    strcpy(tmp, str);
+    pubmsg.payload = tmp;
+    pubmsg.payloadlen = strlen(str);
+    pubmsg.qos = QOS;
+    pubmsg.retained = 0;
+    if ((rc = MQTTAsync_sendMessage(mqtt_client, topic, &pubmsg, NULL)) != MQTTASYNC_SUCCESS) {
+        printf("Failed to start sendMessage, return code %d\n", rc);
+        MQTTHandle_error(rc);
+    }
+}
+
+void AP_Telemetry_MQTT::subscribe_mqtt_topic(const char *topic, int qos)
+{
+    if (connection_status == MQTT_CONNECTED) {
+        int rc;
+        if ((rc = MQTTAsync_subscribe(mqtt_client, topic, qos, NULL)) != MQTTASYNC_SUCCESS) {
+            printf("Failed to start subscribe, return code %d\n", rc);
+            MQTTHandle_error(rc);
+        }
+    }
+}
+
+int AP_Telemetry_MQTT::recv_mavlink_message(mavlink_message_t *msg)
+{
+    int ret = 0;
+    char str_mqtt[MAX_PAYLOAD];
+    str_mqtt[0] = '\0';
+    AP_Telemetry_MQTT::pop_mqtt_message(str_mqtt);
+    if (str_mqtt[0] != '\0') {
+        ret = mqtt_to_mavlink_message(str_mqtt, msg);
+    }
+    return ret;
+}
+
+void AP_Telemetry_MQTT::pop_mqtt_message(char* str_mqtt)
+{
+    MQTTAsync_message* message = nullptr;
+    if (pthread_mutex_lock(AP_Telemetry_MQTT::mqtt_mutex) == 0) {
+        message = (MQTTAsync_message*)ListDetachHead(recv_msg_list);
+        pthread_mutex_unlock(AP_Telemetry_MQTT::mqtt_mutex);
+        if (message != nullptr) {
+            strncpy(str_mqtt, (char*)message->payload, message->payloadlen);
+            str_mqtt[message->payloadlen] = 0;
+        }
+    }
+}
+
+void AP_Telemetry_MQTT::append_mqtt_message(MQTTAsync_message* message)
+{
+    if (pthread_mutex_lock(mqtt_mutex) == 0) {
+        ListAppend(recv_msg_list, message, sizeof(MQTTAsync_message));
+        pthread_mutex_unlock(mqtt_mutex);
+    } else {
+        MQTTAsync_freeMessage(&message);
+    }
+}
+
+void AP_Telemetry_MQTT::MQTTHandle_error(int rc)
+{
+    switch (rc) {
```

Comment:
Is it because the switch statement is planned to be added in the future? If there is no addition, I think that the if statement is good.